### PR TITLE
Add support for batch_norm fusion to the JIT

### PR DESCRIPTION
**Pull request by apaszke**
We don't support reductions yet, but simply decomposing batch_norm
into a kernel that computes the stats, and the fusing everything else
with ReLU and following pointwise ops provides nice speedups.

Note that this is only limited to inference mode for now, because we
don't support convolutions and batch norm in AD, so the fuser isn't
applied to those parts.

This commit gives us a 7% end-to-end speedup for ResNet50 with batch size 32. Note that this only applies to inference mode at the moment due to lack of AD support for CNN operations (I'll be adding that soon), and not to the standard `torchvision` models, because they use in-place ops which aren't supported by the fuser (we need a way of proving that de-inplacing them is safe).

cc @zou3519 @zdevito @mruberry @ngimel

### CODE REVIEW

**Comment by zou3519** [comment]
@apaszke there are some AutodiffSubgraphSlicing tests failing. I'm not sure if those are related to your PR

**Review by zou3519 — COMMENTED** [review]
(still reading through, not a full review yet)

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 1]
```diff
@@ -115,6 +119,45 @@ bool isSimpleMap(Node *node) {
   return true;
 }
 
+RegisterOperators reg_bn_unsqueeze({
+    Operator(
+        "aten::_ncf_expand(Tensor self, int ndim) -> Tensor",
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/passes/graph_fuser.cpp]: nit: _ncf_reshape might be a better name because we are reshaping instead of expanding

File: torch/csrc/jit/fuser/codegen.cpp [hunk 2]
```diff
@@ -134,6 +134,7 @@ static std::string encodeRHS(const Node* n) {
     {aten::abs, "fabs(${0})"},
     {aten::sigmoid, "1.f / (1.f + expf(-${0}))"},
     {aten::relu, "${0} < 0 ? 0.f : ${0} "},
+    {aten::threshold, "${0} < ${1} ? static_cast<decltype(${0})>(${2}) : ${0} "},
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/fuser/codegen.cpp]: We should also probably add a correctness test for this to TestFuser

File: test/test_jit.py [hunk 3]
```diff
@@ -536,6 +536,39 @@ def f(x, y):
...
+            @torch.jit.script_method
+            def forward(self, x, y):
+                return y + torch.relu(self.bn(x))
+
+        model = ResLike().cuda()
+        x = torch.randn(2, 16, 8, 8, device='cuda')
+        y = torch.randn(2, 16, 8, 8, device='cuda')
+        # FIXME: We need differentiation for CNNs for this optimization to trigger
+        with torch.no_grad():
+            model(x, y)
```
Comment (zou3519) [inline test/test_jit.py]: Can we also add a correctness test that runs with the JIT and checks the output?

**Review by zou3519 — COMMENTED** [review]
BN changes look fine, I had some minor questions and comments, please read

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -624,7 +726,7 @@ struct GraphFuser {
         chunk->inputs().begin(),
         chunk->inputs().end(),
         [&](Value * producer_for_chunk) {
-          return isFusable(producer_for_chunk->node()) &&
+          return isFusableMap(producer_for_chunk->node()) &&
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: It wouldn't, but I really don't expect this to be the common case, and I don't want to make the code for chunk more complicated for no good reason.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 5]
```diff
@@ -137,6 +181,10 @@ struct GraphFuser {
   }
 
   bool isFusable(Node * node) {
+    return isFusableMap(node) || isFusableBatchNorm(node);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I agree our definition of "fusable" is completely messed up, but clearing that up is a material for another PR. I simply followed whatever we were using.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 6]
```diff
@@ -199,6 +247,54 @@ struct GraphFuser {
...
+      *graph_ptr = module->get_method("batch_norm").graph();
+    }, &bn_graph);
+
+    JIT_ASSERT(isFusableBatchNorm(batch_norm));
+    WithInsertPoint insert_guard { batch_norm };
+    Value * input = batch_norm->namedInput(attr::input);
+    Value * input_dim = graph_->insert(aten::dim, {input});
+    std::vector<Value*> inputs { input,
+                                 batch_norm->namedInput(attr::running_mean),
+                                 batch_norm->namedInput(attr::running_var),
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Yep, `Optional[Tensor]` means that it can be undefined on the C++ side 😕

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: It's as in NCHW, but it works with any dimension hence F for features instead of HW. I can rename it to reshape if you'd really want to, but I wouldn't like to block this PR on it if that's the only problem.

File: test/test_jit.py [hunk 7]
```diff
@@ -536,6 +536,39 @@ def f(x, y):
 
         self.checkTrace(f, (x, y))
 
+    @unittest.skipIf(not IS_WINDOWS, "Testing Fuse skipped on windows")
+    @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")
+    def test_fuse_batch_norm(self):
```
[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_jit.py]: Yep, will move.

File: torch/csrc/jit/fuser/codegen.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/fuser/codegen.cpp]: It's an orthogonal improvement to the fuser. Lets us fuse whole blocks between convs in ResNets.

File: test/test_jit.py (hunk 3 above)
Comment (apaszke) [reply test/test_jit.py]: Yup, will do.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Yikes, thanks for the clarification

**Review by zou3519 — APPROVED** [review]
lgtm!

**Comment by zou3519** [comment]
@apaszke looks like you might have to rebase this

**Comment by apaszke** [comment]
ROCm failures look unrelated

**Comment by apaszke** [comment]
I forgot to add CPU implementation for `batch_norm_update_stats`, which is fixed in the latest commit. Please review the ATen changes again!

**Comment by apaszke** [comment]
Red jobs are CI failures.

File: aten/src/ATen/native/Normalization.cpp [hunk 8]
```diff
@@ -36,83 +36,116 @@ static TensorAccessor<scalar_t, 1> conditional_accessor_1d(const Tensor& t) {
...
+};
 
 template<typename scalar_t>
-std::tuple<Tensor,Tensor,Tensor> batch_norm_cpu_template(const Tensor& input, const Tensor& weight, const Tensor& bias,
-                               const Tensor& running_mean, const Tensor& running_var, bool train, double momentum, double eps) {
+std::tuple<Tensor,Tensor,Tensor> batch_norm_cpu_transform_input_template(
+    const Tensor& input, const Tensor& weight, const Tensor& bias,
+    const Tensor& save_mean /* optional */, const Tensor& save_invstd /* optional */,
+    const Tensor& running_mean /* optional */, const Tensor& running_var /* optional */,
+    bool train, double momentum, double eps) {
```
Comment (zou3519) [inline aten/src/ATen/native/Normalization.cpp]: `momentum` is unused in this function; it's only necessary for updating the running stats

File: aten/src/ATen/native/Normalization.cpp [hunk 9]
```diff
@@ -36,83 +36,116 @@ static TensorAccessor<scalar_t, 1> conditional_accessor_1d(const Tensor& t) {
...
 
-  using accscalar_t = at::acc_type<scalar_t, false>;
   Tensor output = at::empty_like(input);
 
   int64_t n_input = input.size(1);
   int64_t n = input.numel() / n_input;
```
Comment (zou3519) [inline aten/src/ATen/native/Normalization.cpp]: This is unused now

File: aten/src/ATen/native/Normalization.cpp [hunk 10]
```diff
@@ -36,83 +36,116 @@ static TensorAccessor<scalar_t, 1> conditional_accessor_1d(const Tensor& t) {
...
+    if (var != static_cast<T>(0) || epsilon != static_cast<T>(0)) {
+      invstd = static_cast<T>(1) / std::sqrt(var + epsilon);
+    }
+    return invstd;
+  }
+};
+
+template<typename T>
+struct Var {
+  T operator()(T var, double epsilon) const {
```
Comment (zou3519) [inline aten/src/ATen/native/Normalization.cpp]: Nit: epsilon not used, maybe remove the variable name?

**Review by zou3519 — APPROVED** [review]
LGTM. There are some unused variables, let me know if you want to land this as-is or if you want to clean them up first

**Comment by zou3519** [comment]
@apaszke this needs a rebase now

**Comment by chanil1218** [comment]
@apaszke 
Thank you for releasing nice code!
What I understood is that JIT fused batch_norm should only be used at inference time for speed up purpose. Due to the reason that batch norm is not supported in AD.

Is there any update on **AD support for batch norm** for using fused batch_norm on training time? 
Or related issue that I could keep track of the progress of that support?
